ncm-opennebula/src/main/perl/OpenNebula/Host.pm

```diff
@@ -0,0 +1,169 @@
+        } elsif (exists($newhosts{$t->name})) {
+            $self->debug(1, "We can't remove this $type host. Is required by Quattor: ", $t->name);
+        } elsif ($t->used()) {
+            $self->debug(1, "We can't remove this $type host. There are still running VMs: ", $t->name);
+        } else {
+            push(@rmhosts, $t->name);
+            $t->delete();
+        }
+    }
+
+    if (@rmhosts) {
+        $self->info("Removed $type hosts: ", join(',', @rmhosts));
+    }
+
+    foreach my $host (@$hosts) {
+        my %host_options = (
+            'name'    => $host, 
+            'im_mad'  => $type, 
+            'vmm_mad' => $type, 
+            'vnm_mad' => $vnm_mad
+        );
+        # to keep the record of our cloud infrastructure
+        # we include the host in ONE db even if it fails
+        my @hostinstances = $one->get_hosts(qr{^$host$});
+        if (scalar(@hostinstances) > 1) {
+            $self->error("Found more than one host $host. Only the first host will be modified.");
+        }
+        my $hostinstance = $hostinstances[0];
+        if (exists($protected{$host})) {
+            $self->info("This resource $type is protected and can not be created/updated: $host");
+        } elsif ($self->can_connect_to_host($host)) {
+            my $output = $self->enable_node($one, $type, $host, $resources);
+            if ($output) {
+                if ($hostinstance) {
+                    # The host is already available and OK
+                    $hostinstance->enable if !$hostinstance->is_enabled;
+                    $self->info("Enabled existing host $host");
+                } else {
+                    # The host is not available yet from ONE framework
+                    # and it is running correctly
+                    $new = $one->create_host(%host_options);
+                    $self->update_something($one, "host", $host, "QUATTOR = 1");
+                    $self->info("Created new $type host $host.");
+                }
+            } else {
+                $self->
+                disable_host($one, $type, $host, $hostinstance, %host_options);
+            }
+        } else {
+            push(@failedhost, $host);
+            $self->
+            disable_host($one, $type, $host, $hostinstance, %host_options);
+        }
+    }
+
+    if (@failedhost) {
+        $self->error("Detected some error/s including these $type nodes: ", join(', ', @failedhost));
+    }
+
+}
+
+=item disable_host
+
+Disables failing OpenNebula C<host>.
+
+=cut
+
+sub disable_host
+{
+    my ($self, $one, $type, $host, $hostinstance, %host_options) = @_;
+
+    $self->warn("Could not connect to $type host: $host.");
+    if ($hostinstance) {
+        $hostinstance->disable;
+        $self->info("Disabled existing host $host");
+    } else {
+        my $new = $one->create_host(%host_options);
+        $new->disable;
+        $self->info("Created and disabled new host $host");
+    }
+}
+
+=item sync_opennebula_hyps
```

Comment:
yes is a mess I will use host/hosts because is the name chosen by OpenNebula commands